extensions/amp-animation/0.1/keyframes-extractor.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @extends {CSSRule}
+ * @see http://www.w3.org/TR/DOM-Level-2-Style/css.html#CSS-CSSKeyframesRule
+ */
+class CSSKeyframesRuleDef {
+  constructor() {
+    /** @type {string} */
+    this.name;
+    /** @type {string} */
+    this.keyText;
+    /** @type {!CSSRuleList} */
+    this.cssRules;
+  }
+}
+
+
+/**
+ * Finds and extracts keyframes definition for Web Animations from CSS styles.
+ * @param {!Document|!ShadowRoot} rootNode
+ * @param {string} name
+ * @return {?./web-animation-types.WebKeyframesDef}
+ */
+export function extractKeyframes(rootNode, name) {
+  const styleSheets = rootNode.styleSheets;
+  if (!styleSheets) {
+    return null;
+  }
+  const win = (rootNode.ownerDocument || rootNode).defaultView;
+  // Go from the last to first since the last rule wins in CSS.
+  for (let i = styleSheets.length - 1; i >= 0; i--) {
+    const keyframes = scanStyle(
+        win, /** @type {!CSSStyleSheet} */ (styleSheets[i]), name);
+    if (keyframes) {
+      return keyframes;
+    }
+  }
+  return null;
+}
+
+
+/**
+ * @param {!Window} win
+ * @param {!CSSStyleSheet} styleSheet
+ * @param {string} name
+ * @return {?./web-animation-types.WebKeyframesDef}
+ */
+function scanStyle(win, styleSheet, name) {
+  // No rules, e.g. a font.
+  if (!styleSheet.cssRules) {
+    return null;
+  }
+
+  const styleNode = styleSheet.ownerNode;
+  if (!styleNode) {
+    return null;
+  }
+  // Exlcude AMP's own styles.
+  if (styleNode.hasAttribute('amp-boilerplate') ||
+      styleNode.hasAttribute('amp-runtime') ||
+      styleNode.hasAttribute('amp-extension')) {
+    return null;
+  }
+
+  return scanRules(win, styleSheet.cssRules, name);
+}
+
+
+/**
+ * @param {!Window} win
+ * @param {!CSSRuleList} rules
+ * @param {string} name
+ * @return {?./web-animation-types.WebKeyframesDef}
+ */
+function scanRules(win, rules, name) {
+  // Go backwards since in CSS the last one wins.
+  for (let i = rules.length - 1; i >= 0; i--) {
+    const rule = rules[i];
+    if (rule.type == /* CSSKeyframesRule */ 7) {
+      const keyframesRule = /** @type {!CSSKeyframesRuleDef} */ (rule);
+      if (rule.name == name && isEnabled(win, rule)) {
+        return buildKeyframes(keyframesRule);
+      }
+    } else if (rule.type == /* CSSMediaRule */ 4) {
```

Comment:
The reason I didn't do this is because there are potentially a lot more `@media` definitions than `@keyframes`. And checking in each case is somewhat expensive. So instead I first find a likely candidate `@keyframes name` and then check if its parents are enabled recursively.